### PR TITLE
add benchmark and check program for core ngt

### PR DESCRIPTION
**Pull request by kmrmt**
### Description:

Implement core algorithm NGT's benchmark and getting metrics about memory usage.
It has 2 way of measuring memory usage.
One way is to use cmd/tools/cli/benchmark/core/main.go while another way is to use internal/core/algorithm/ngt/ngt_bench_test.go.
If you want detailed metrics, I recommend you to use the first way.
And include some fix.

### Versions:

- Go Version: 1.21.1
- Docker Version: 20.10.8
- Kubernetes Version: v1.28.1
- NGT Version: 2.1.3

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/main/CONTRIBUTING.md) document and completed [our CLA agreement](https://cla-assistant.io/vdaas/vald).
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?

### CODE REVIEW

File: internal/net/grpc/pool/pool.go [hunk 1]
```diff
@@ -95,31 +94,35 @@ func New(ctx context.Context, opts ...Option) (c Conn, err error) {
...
 		}
 		if p.port == 0 {
-			err = p.scanGRPCPort(ctx)
+			port, err = p.scanGRPCPort(ctx)
 			if err != nil {
 				return nil, err
 			}
+			p.port = port
 		}
 		p.addr = net.JoinHostPort(p.host, p.port)
```
[the remark this answers is not in the document]
Comment (kmrmt) [reply internal/net/grpc/pool/pool.go:127]: kpango changes here.
@kpango Could you please answer to this comment?

File: internal/core/algorithm/ngt/ngt_bench_test.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+	n, _ = New(
+		WithDimension(len(vectors[0])),
+		WithDefaultPoolSize(8),
+		WithObjectType(Float),
+		WithDistanceType(L2),
+	)
+	pid = os.Getpid()
+}
+
+func BenchmarkNGT(b *testing.B) {
```
[the remark this answers is not in the document]
Comment (kmrmt) [reply internal/core/algorithm/ngt/ngt_bench_test.go:52]: Yes, I added description.

File: cmd/tools/cli/benchmark/core/main.go [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"runtime"
+	"strconv"
+	"sync"
+	"time"
```
[the remark this answers is not in the document]
Comment (kmrmt) [reply cmd/tools/cli/benchmark/core/main.go]: fix it.

**Comment by ykadowak** [comment]
I'll approve once the conflicts are resolved

**Review by kpango — APPROVED** [review]
LGTM

**Pull request merged**